### ISSUE TITLE
datastore: performance degradation migrating to App Engine Go1.11

### ISSUE THREAD

**Issue by ericlevine** [googleapis/google-cloud-go#1295]
## Client

Datastore

## Describe Your Environment

Go 1.11 runtime on Google App Engine 2nd generation

## Expected Behavior

Datastore has similar performance to the App Engine datastore client.

## Actual Behavior

Significant performance degradation from the App Engine datastore client.

I'm seeing 100-200ms key lookups on datastore entities that are not very large. This has caused a pretty serious performance regression for my application. This appears to have been an issue [in the past](https://github.com/googleapis/google-cloud-go/issues/135), but that it was supposedly addressed by adding gRPC support.

Now that the Go 1.11 on App Engine migration guide suggests using the cloud datastore package, I suspect this is going to be a recurring issue for apps trying to migrate.

Are there any client configurations that could bring performance to parity with the App Engine package, or is the best path forward to keep using the App Engine package for the foreseeable future?

**Comment by jadekler** [comment]
cc @tbpg @andrewsg

**Comment by ericlevine** [comment]
For additional context, after reverting to using the App Engine client, typical overall response times are 1/2 to 1/3 what they are using the other client with no other changes other than what's required to use the App Engine client.

**Comment by sbuss** [comment]
@ericlevine What region are you running in? There is an ongoing push to fix a slow networking path in the SE Asia area -- are you running your code there by chance?

**Comment by ericlevine** [comment]
@sbuss We're running in us-central1. The performance issues aren't critical for us anymore given that we reverted to the App Engine client for the time being.

**Comment by sbuss** [comment]
I've opened an internal bug to investigate this. Will update when I have info

**Comment by zaddok** [comment]
For what it’s worth, the average speed of fetching small entities in the AUS region has just dropped from ~850ms down to ~15ms. Very happy about this—850ms was for all practical purposes essentially broken.

**Comment by jadekler** [comment]
Friendly ping on this thread.

@sbuss any more to share?

@ericlevine are you still seeing this issue?

**Comment by ericlevine** [comment]
@jadekler We reverted to the App Engine client and haven't bothered trying to switch back to the cloud datastore package in the absence of any updates on this issue.

**Comment by jadekler** [comment]
Ack, thanks. @sbuss do you want me to keep this open for tracking purposes or can I close this/

**Comment by ericlevine** [comment]
Just revisiting this. I came up with a minimal go app that exhibits severe performance issues. It's unclear to me if there are some easy client option changes that are required to get reasonable performance, but please see this gist:

https://gist.github.com/ericlevine/27f491ef3cf65406a29b785be89d5a69

When this minimal go 1.11 app engine app is run in us-central1:

- `put`s take 100ms to 400ms, averaging around 150ms
- `get`s take 30ms to 100ms, averaging around 50ms

Is this expected performance? Does the App Engine Datastore client perform memcache operations opaquely to improve performance?

**Comment by sbuss** [comment]
Sorry for the delayed response; these notifications were going to my personal email instead of my @google.com email.

@ericlevine what region are you running in, and what region is your datastore? AFAIK there's still an issue with inter-region latencies (but I haven't looked in a while).

**Comment by ericlevine** [comment]
No worries, thanks for the quick ack, @sbuss. Both should be us-central, though all the documentation that I've found on precise region for datastore implies that it's the project region, which seems to be the same region as App Engine. If there's a better way to get the precise locations, happy to track those down!

**Comment by andwun** [comment]
@ericlevine I saw from your gist that you do not reuse the datastore client across requests.
tldr: Try setting up the datastore client once in main and reuse it in all your handlers.

Background as I see it:
As far as I understand the situation, the old appengine datastore client uses special RPCs from the appengine environment.
The new datastore client instead uses HTTPS for communication with the datastore service.
If you do not reuse the datastore client across requests, it will not reuse the underlying HTTPS connections, which means it has to establish a new HTTPS connection on each request coming into your appengine app.
This can take a lot of CPU cycles, causes high latency and should thus be avoided.

**Comment by ericlevine** [comment]
Hey @andwun, thanks for the response. I gave that a shot and updated the gist to reflect the change you suggested so that it creates one client on demand and keeps that in memory.

https://gist.github.com/ericlevine/27f491ef3cf65406a29b785be89d5a69

Here are the results:

```
2019/07/29 15:51:38 makeClient took 608ns
 2019/07/29 15:51:38 put took 159.2569ms
 2019/07/29 15:51:38 get took 135.396093ms
 2019/07/29 15:51:38 put took 291.8579ms
 2019/07/29 15:51:38 get took 134.992958ms
 2019/07/29 15:51:39 put took 231.121845ms
 2019/07/29 15:51:39 get took 44.22318ms
 2019/07/29 15:51:39 request took 997.169022ms
 "GET / HTTP/1.1" 200
 2019/07/29 15:51:43 makeClient took 602ns
 2019/07/29 15:51:43 put took 136.263303ms
 2019/07/29 15:51:44 get took 124.116754ms
 2019/07/29 15:51:44 put took 71.096918ms
 2019/07/29 15:51:44 get took 32.436369ms
 2019/07/29 15:51:44 put took 96.37074ms
 2019/07/29 15:51:44 get took 74.811275ms
 2019/07/29 15:51:44 request took 535.502469ms
 "GET / HTTP/1.1" 200
 2019/07/29 15:51:48 makeClient took 492ns
 2019/07/29 15:51:48 put took 123.530844ms
 2019/07/29 15:51:48 get took 126.828645ms
 2019/07/29 15:51:49 put took 204.167607ms
 2019/07/29 15:51:49 get took 50.95261ms
 2019/07/29 15:51:49 put took 217.027777ms
 2019/07/29 15:51:49 get took 103.920772ms
 2019/07/29 15:51:49 request took 826.645039ms
```

Are there any special options that need to be provided to the datastore client to ensure that they're reusing the existing HTTPS connection? It certainly sounds like it could be the issue.

**Comment by andwun** [comment]
@ericlevine We setup our datastore client with GRPC options found from the pubsub client:
https://github.com/googleapis/google-cloud-go/blob/d86525d39e84b290a479921b13c3b366df35d9fd/pubsub/pubsub.go#L67

This should enable connection pooling and connection keep alive.
Hope that improves the situation for you, too.

**Comment by ericlevine** [comment]
@andwun This certainly seems to have improved the performance significantly. I believe this is still underperforming the old AppEngine client, but the improvement is definitely non-negligible.

```
2019/07/29 18:26:09 makeClient took 357ns
2019/07/29 18:26:10 put took 167.709391ms
2019/07/29 18:26:10 get took 39.150635ms
2019/07/29 18:26:10 put took 122.892409ms
2019/07/29 18:26:10 get took 41.439258ms
2019/07/29 18:26:10 put took 166.51365ms
2019/07/29 18:26:10 get took 53.048786ms
2019/07/29 18:26:10 request took 590.967825ms

2019/07/29 18:26:14 makeClient took 334ns
2019/07/29 18:26:15 put took 137.395257ms
2019/07/29 18:26:15 get took 85.799839ms
2019/07/29 18:26:15 put took 107.1005ms
2019/07/29 18:26:15 get took 145.138178ms
2019/07/29 18:26:15 put took 140.137944ms
2019/07/29 18:26:15 get took 82.216126ms
2019/07/29 18:26:15 request took 698.005427ms

2019/07/29 18:26:20 makeClient took 322ns
2019/07/29 18:26:20 put took 158.619154ms
2019/07/29 18:26:20 get took 41.105095ms
2019/07/29 18:26:20 put took 204.736753ms
2019/07/29 18:26:20 get took 36.520354ms
2019/07/29 18:26:20 put took 142.159139ms
2019/07/29 18:26:20 get took 37.435879ms
2019/07/29 18:26:20 request took 620.838762ms
```

As a separate note, the documentation seems to imply that Datastore is still using gRPC (as it's not one of the callouts for HTTP).

**Comment by brugz** [comment]
FWIW, I think some of this behavior may also observable in golang/appengine/datastore on golang 1.19. I will open an issue against that repo when I have some more information.

https://github.com/golang/appengine/tree/master/datastore

**Comment by gleno** [comment]
Still observing performance loss. 

The above suggestions about keep-alives seem to help. 

Any updates?

**Comment by MeganSpencer** [comment]
The latency is oddly slow with my app in the Australian region. Not a fix, but my workaround is to use go routines. Do your 6 operations have to be sequential?

Here is my workaround:

                var wg sync.WaitGroup
                wg.Add(3)
                go func() {
                        defer wg.Done()
                        user, err1 = ss.GetUserCached(uuid, session)
                }()
                go func() {
                        defer wg.Done()
                        // do stuff
                }()
                go func() {
                        defer wg.Done()
                        // do stuff
                }()
                wg.Wait()

                if err1 != nil {
                        security.ShowError(w, r, t, err1, session)
                        return
                }

**Comment by crwilcox** [comment]
I am going to close this issue out. There are performance differences with appengine datastore and datastore to date. In 
 reviewing this issue the outstanding bits seem to be more questions of the service performance than the client. Datastore service performance will see improvements over time but this repo really doesn't track that part of the effort. Thanks to everyone for the feedback and help making improvements here and there to our client libraries.
